corporate/tests/test_stripe.py

```diff
@@ -5432,3 +5437,98 @@ def test_downgrade_realm_and_void_open_invoices(self, *mocks: Mock) -> None:
         self.assertEqual(success_message, "zulip downgraded and voided 1 open invoices")
         original_plan.refresh_from_db()
         self.assertEqual(original_plan.status, CustomerPlan.ENDED)
+
+
+class TestRemoteBillingWriteAuditLog(StripeTestCase):
+    def test_write_audit_log(self) -> None:
+        support_admin = self.example_user("iago")
+        server_uuid = str(uuid.uuid4())
+        remote_server = RemoteZulipServer.objects.create(
+            uuid=server_uuid,
+            api_key="magic_secret_api_key",
+            hostname="demo.example.com",
+            contact_email="email@example.com",
+        )
+        realm_uuid = str(uuid.uuid4())
+        remote_realm = RemoteRealm.objects.create(
+            server=remote_server,
+            uuid=realm_uuid,
+            uuid_owner_secret="dummy-owner-secret",
+            host="dummy-hostname",
+            realm_date_created=timezone_now(),
+        )
+        remote_realm_billing_user = RemoteRealmBillingUser.objects.create(
+            remote_realm=remote_realm, email="admin@example.com", user_uuid=uuid.uuid4()
+        )
+        remote_server_billing_user = RemoteServerBillingUser.objects.create(
+            remote_server=remote_server, email="admin@example.com"
+        )
+        event_time = timezone_now()
+
+        def assert_audit_log(
+            audit_log: Union[RemoteRealmAuditLog, RemoteZulipServerAuditLog],
+            acting_remote_user: Optional[Union[RemoteRealmBillingUser, RemoteServerBillingUser]],
+            acting_support_user: Optional[UserProfile],
+            event_type: int,
+            event_time: datetime,
+        ) -> None:
+            self.assertEqual(audit_log.event_type, event_type)
+            self.assertEqual(audit_log.event_time, event_time)
+            self.assertEqual(audit_log.acting_remote_user, acting_remote_user)
+            self.assertEqual(audit_log.acting_support_user, acting_support_user)
+
+        for session_class, audit_log_class, remote_object, remote_user in [
+            (
+                RemoteRealmBillingSession,
+                RemoteRealmAuditLog,
+                remote_realm,
+                remote_realm_billing_user,
+            ),
+            (
+                RemoteServerBillingSession,
+                RemoteZulipServerAuditLog,
+                remote_server,
+                remote_server_billing_user,
+            ),
+        ]:
+            # Necessary cast or mypy doesn't understand that we can use Django's
+            # model .objects. style queries on this.
+            audit_log_model = cast(
+                Union[Type[RemoteRealmAuditLog], Type[RemoteZulipServerAuditLog]], audit_log_class
+            )
```

Comment:
Mypy hates me lately :confused: If I do this without casts, and just using `audit_log_class`, which I think it should be able to correctly infer, it throws
```
error: "Type[AbstractRealmAuditLog]" has no attribute "objects"
```
errors